File: benchmarks/float_compute.rs
use std::hint::black_box;
use std::time::Instant;

const RUNS: usize = 3;

fn sum_inv_squares(n: i64) -> f64 {
    let mut sum: f64 = 0.0;
    for i in 1..=n {
        let fi = i as f64;
        sum += 1.0 / (fi * fi);
    }
    sum
}

fn leibniz_pi(n: i64) -> f64 {
    let mut sum: f64 = 0.0;
    let mut sign: f64 = 1.0;
    for i in 1..=n {
        let denom = (i * 2 - 1) as f64;
        sum += sign / denom;
        sign = -sign;
    }
    sum * 4.0
}

fn mandelbrot_iter(cr: f64, ci: f64, max_iter: i64) -> i64 {
    let mut zr: f64 = 0.0;
    let mut zi: f64 = 0.0;
    for i in 0..max_iter {
        let tr = zr * zr - zi * zi + cr;
        zi = 2.0 * zr * zi + ci;
        zr = tr;
        if zr * zr + zi * zi > 4.0 {
            return i;
        }
    }
    max_iter
}

fn distance(x1: f64, y1: f64, x2: f64, y2: f64) -> f64 {
    let dx = x2 - x1;
    let dy = y2 - y1;
    (dx * dx + dy * dy).sqrt()
}

fn trig_sum(n: i64) -> f64 {
    let mut sum: f64 = 0.0;
    for i in 0..n {
        let x = i as f64 * 0.001;
        sum += x.sin() + x.cos();
    }
    sum
}

fn main() {
    println!("=== Rust Float Benchmarks ===");

    // Benchmark 1: Basel problem
    let n: i64 = 10000000;
    let _ = sum_inv_squares(black_box(n)); // warmup
    let mut best: u128 = u128::MAX;
    let mut result: f64 = 0.0;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = sum_inv_squares(black_box(n));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Basel sum(1/i^2, 1..{}):", n);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {:.10}", result);

    // Benchmark 2: Leibniz pi
    let n: i64 = 10000000;
    let _ = leibniz_pi(black_box(n)); // warmup
    best = u128::MAX;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = leibniz_pi(black_box(n));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Leibniz pi({} terms):", n);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {:.10}", result);

    // Benchmark 3: Mandelbrot grid
    let size: i64 = 200;
    // warmup
    {
        let mut t: i64 = 0;
        for y in 0..size {
            for x in 0..size {
                let cr = (x as f64 - 100.0) / 50.0;
                let ci = (y as f64 - 100.0) / 50.0;
                t += mandelbrot_iter(cr, ci, 100);
            }
        }
        let _ = t;
    }
    best = u128::MAX;
    let mut total: i64 = 0;
    for _ in 0..RUNS {
        let start = Instant::now();
        let mut t: i64 = 0;
        for y in 0..size {
            for x in 0..size {
                let cr = (x as f64 - 100.0) / 50.0;
                let ci = (y as f64 - 100.0) / 50.0;
                t += mandelbrot_iter(cr, ci, 100);
            }
        }
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; total = t; }
    }
    println!("Mandelbrot {}x{}:", size, size);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {}", total);

    // Benchmark 4: Distance computation
    let n: i64 = 10000000;
    // warmup
    {
        let mut s: f64 = 0.0;
        for i in 0..n {
            let x = i as f64 * 0.001;
            s += distance(black_box(x), black_box(0.0), black_box(0.0), black_box(x));
        }
        let _ = s;
    }
    best = u128::MAX;
    for _ in 0..RUNS {
        let start = Instant::now();
        let mut s: f64 = 0.0;
        for i in 0..n {
            let x = i as f64 * 0.001;
            s += distance(black_box(x), black_box(0.0), black_box(0.0), black_box(x));
        }
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = s; }
    }
    println!("Distance ({} iterations):", n);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {:.10}", result);

    // Benchmark 5: Trig computation
    let n: i64 = 5000000;
    let _ = trig_sum(black_box(n)); // warmup
    best = u128::MAX;
    for _ in 0..RUNS {
        let start = Instant::now();
        let r = trig_sum(black_box(n));
        let elapsed = start.elapsed().as_nanos();
        if elapsed < best { best = elapsed; result = r; }
    }
    println!("Trig sum({}):", n);
    println!("  Time: {} ms", best / 1000000);
    println!("  Result: {:.10}", result);
}

File: benchmarks/float_compute.cs
using System.Diagnostics;

const int Runs = 3;

double SumInvSquares(long n) {
    double sum = 0.0;
    for (long i = 1; i <= n; i++) {
        double fi = (double)i;
        sum += 1.0 / (fi * fi);
    }
    return sum;
}

double LeibnizPi(long n) {
    double sum = 0.0;
    double sign = 1.0;
    for (long i = 1; i <= n; i++) {
        double denom = (double)(i * 2 - 1);
        sum += sign / denom;
        sign = -sign;
    }
    return sum * 4.0;
}

double Distance(double x1, double y1, double x2, double y2) {
    double dx = x2 - x1;
    double dy = y2 - y1;
    return Math.Sqrt(dx * dx + dy * dy);
}

double TrigSum(long n) {
    double sum = 0.0;
    for (long i = 0; i < n; i++) {
        double x = (double)i * 0.001;
        sum += Math.Sin(x) + Math.Cos(x);
    }
    return sum;
}

long MandelbrotIter(double cr, double ci, long maxIter) {
    double zr = 0.0, zi = 0.0;
    for (long i = 0; i < maxIter; i++) {
        double tr = zr * zr - zi * zi + cr;
        zi = 2.0 * zr * zi + ci;
        zr = tr;
        if (zr * zr + zi * zi > 4.0) {
            return i;
        }
    }
    return maxIter;
}

Console.WriteLine("=== C# Float Benchmarks ===");

// Benchmark 1: Basel problem
long n = 10000000;
SumInvSquares(n); // warmup
long best = long.MaxValue;
double resultF = 0.0;
for (int run = 0; run < Runs; run++) {
    var sw = Stopwatch.StartNew();
    double r = SumInvSquares(n);
    sw.Stop();
    long elapsed = sw.Elapsed.Ticks * 100;
    if (elapsed < best) { best = elapsed; resultF = r; }
}
Console.WriteLine($"Basel sum(1/i^2, 1..{n}):");
Console.WriteLine($"  Time: {best / 1000000} ms");
Console.WriteLine($"  Result: {resultF:F10}");

// Benchmark 2: Leibniz pi
n = 10000000;
LeibnizPi(n); // warmup
best = long.MaxValue;
for (int run = 0; run < Runs; run++) {
    var sw = Stopwatch.StartNew();
    double r = LeibnizPi(n);
    sw.Stop();
    long elapsed = sw.Elapsed.Ticks * 100;
    if (elapsed < best) { best = elapsed; resultF = r; }
}
Console.WriteLine($"Leibniz pi({n} terms):");
Console.WriteLine($"  Time: {best / 1000000} ms");
Console.WriteLine($"  Result: {resultF:F10}");

// Benchmark 3: Mandelbrot grid
long size = 200;
// warmup
{
    long t = 0;
    for (long y = 0; y < size; y++)
        for (long x = 0; x < size; x++)
            t += MandelbrotIter(((double)x - 100.0) / 50.0, ((double)y - 100.0) / 50.0, 100);
}
best = long.MaxValue;
long total = 0;
for (int run = 0; run < Runs; run++) {
    var sw = Stopwatch.StartNew();
    long t = 0;
    for (long y = 0; y < size; y++) {
        for (long x = 0; x < size; x++) {
            double cr = ((double)x - 100.0) / 50.0;
            double ci = ((double)y - 100.0) / 50.0;
            t += MandelbrotIter(cr, ci, 100);
        }
    }
    sw.Stop();
    long elapsed = sw.Elapsed.Ticks * 100;
    if (elapsed < best) { best = elapsed; total = t; }
}
Console.WriteLine($"Mandelbrot {size}x{size}:");
Console.WriteLine($"  Time: {best / 1000000} ms");
Console.WriteLine($"  Result: {total}");

// Benchmark 4: Distance computation
n = 10000000;
// warmup
{
    double s = 0.0;
    for (long i = 0; i < n; i++) {
        double x = (double)i * 0.001;
        s += Distance(x, 0.0, 0.0, x);
    }
}
best = long.MaxValue;
for (int run = 0; run < Runs; run++) {
    var sw = Stopwatch.StartNew();
    double s = 0.0;
    for (long i = 0; i < n; i++) {
        double x = (double)i * 0.001;
        s += Distance(x, 0.0, 0.0, x);
    }
    sw.Stop();
    long elapsed = sw.Elapsed.Ticks * 100;
    if (elapsed < best) { best = elapsed; resultF = s; }
}
Console.WriteLine($"Distance ({n} iterations):");
Console.WriteLine($"  Time: {best / 1000000} ms");
Console.WriteLine($"  Result: {resultF:F10}");

// Benchmark 5: Trig computation
n = 5000000;
TrigSum(n); // warmup
best = long.MaxValue;
for (int run = 0; run < Runs; run++) {
    var sw = Stopwatch.StartNew();
    double r = TrigSum(n);
    sw.Stop();
    long elapsed = sw.Elapsed.Ticks * 100;
    if (elapsed < best) { best = elapsed; resultF = r; }
}
Console.WriteLine($"Trig sum({n}):");
Console.WriteLine($"  Time: {best / 1000000} ms");
Console.WriteLine($"  Result: {resultF:F10}");
